### ISSUE TITLE
ngx.req.raw_header segfault with malformed request

### ISSUE THREAD

**Issue by GordonMcKinney** [openresty/lua-nginx-module#2001]
The following segfault was observed in production with nginx openresty/1.19.3.1:
 
```
Jan 19 19:21:12 kong-enterprise-data-i-0f71bb0913edb5483 kernel: nginx[9244]:
segfault at 0 ip 00000000004dc87a sp 00007ffc390a0c80 error 4 in nginx[400000+1ea000]
```
 
Which resolves to: https://github.com/openresty/lua-nginx-module/blob/v0.10.19/src/ngx_http_lua_headers.c#L170

We distilled the configuration and test cases to a bare-bones OpenResty/1.19.3.1 distribution (no Kong code).

nginx.conf

```
#user  nobody;
worker_processes  1;

#error_log  logs/error.log;
#error_log  logs/error.log  notice;
#error_log  logs/error.log  info;

#pid        logs/nginx.pid;


events {
    worker_connections  1024;
}


http {
    lua_shared_dict log_dict 5M;
    include       mime.types;
    default_type  application/octet-stream;

    log_format  main  '$remote_addr - $remote_user [$time_local] "$request" '
                      '$status $body_bytes_sent "$http_referer" '
                      '"$http_user_agent" "$http_x_forwarded_for" '
                      'rhs=$raw_header_size';

    access_log  logs/access.log  main;

    sendfile        on;
    #tcp_nopush     on;

    #keepalive_timeout  0;
    keepalive_timeout  65;

    #gzip  on;

    server {
        listen       8080;
        server_name  localhost;

        #charset koi8-r;

        #access_log  logs/host.access.log  main;

        location / {
            set    $raw_header_size 0;
            root   html;
            index  index.html index.htm;
        }

        #error_page  404              /404.html;

        # redirect server error pages to the static page /50x.html
        #
        error_page   500 502 503 504  /50x.html;
        location = /50x.html {
            root   html;
        }

        log_by_lua_block {
            ngx.var.raw_header_size = (ngx.req.http_version() < 2.0 and string.len(ngx.req.raw_header())) or 0
        }
    }
}
```

We can trigger the segfault with the following commands:

```
echo "" | nc localhost 8080
echo "qwerty" | nc localhost 8080
echo "GET / HTTP/9" | nc localhost 8080
echo "GET / HTTP\nHost: localhost\n" | nc localhost 8080
```

NOTE: Enabling debug logging prevents the segfault.

Please reach out if you have questions! :-)

**Comment by zhuizhuhaomeng** [comment]
Thanks for your report!
Would you please create a PR for this issue @GordonMcKinney

**Comment by GordonMcKinney** [comment]
@lijunlong Thanks for the prompt review and response! :-)

I would love to create a PR, but my company requires a legal review before submitting any OSS PRs that stem from my day job.

I will start that process unless you see a quicker way to merge this fix into OpenResty 1.19.3.x ([change log](https://openresty.org/en/changelog-1019003.html)). We use Kong in production, and 1.19.3.x is the baseline for their current release train.

TIA :-)

**Comment by GordonMcKinney** [comment]
@zhuizhuhaomeng PR created! LMK if I need to provide anything else.

Of note, we need this bug fix in the 1.19.3.x distribution for Kong to consume it.

**Comment by zhuizhuhaomeng** [comment]
We need to add a test case and make sure the test case can reproduce the crash.
If you want to use the 1.19.3.x, you need to build it yourself

**Comment by zhuizhuhaomeng** [comment]
```
commit c38c470ed16ad29bd91046a0ce90f1369baea80a
Author: lijunlong <lijunlong@openresty.com>
Date:   Sun Feb 6 19:47:36 2022 +0800

    return empty string on invalid request.

diff --git a/src/ngx_http_lua_headers.c b/src/ngx_http_lua_headers.c
index b8eb8b38..54977dce 100644
--- a/src/ngx_http_lua_headers.c
+++ b/src/ngx_http_lua_headers.c
@@ -167,9 +167,13 @@ ngx_http_lua_ngx_req_raw_header(lua_State *L)
     size = 0;
     b = c->buffer;
 
-    if (mr->request_line.len > 0
-        && mr->request_line.data[mr->request_line.len] == CR)
-    {
+    if (mr->request_line.len == 0) {
+        /* return empty string on invalid request */
+        lua_pushlstring(L, "", 0);
+        return 1;
+    }
+
+    if (mr->request_line.data[mr->request_line.len] == CR) {
         line_break_len = 2;
 
     } else {
diff --git a/t/104-req-raw-header.t b/t/104-req-raw-header.t
index f147b2bc..aa666304 100644
--- a/t/104-req-raw-header.t
+++ b/t/104-req-raw-header.t
@@ -1017,3 +1017,36 @@ client sent invalid header line: "\x20..." while reading client request headers
 [error]
 --- skip_nginx
 3: < 1.21.1
+
+
+
+=== TEST 35: bugfix: invalid http request
+--- log_level: error
+--- http_config
+    lua_package_path "../lua-resty-core/lib/?.lua;;";
+--- config
+    location /t {
+        content_by_lua_block {
+            local sock = ngx.socket.tcp()
+            local ok, err = sock:connect("127.0.0.1", $TEST_NGINX_SERVER_PORT)
+            if not ok then
+                ngx.log(ngx.ERR, "failed to connect to memc: ", err)
+                return
+            end
+            sock:send("\n")
+            sock:close()
+
+            ngx.say("OK")
+        }
+    }
+
+    log_by_lua_block {
+        local h = ngx.req.raw_header()
+    }
+
+--- request
+GET /t
+--- response_body
+OK
+--- no_error_log
+[error]
```

@GordonMcKinney  here is my patch. Would you please apply it?

**Comment by GordonMcKinney** [comment]
@zhuizhuhaomeng I will integrate this patch and test it locally. I will contact our OSS program office to clarify whether _your_ patch needs to be approved before I push it.

**Comment by GordonMcKinney** [comment]
@zhuizhuhaomeng findings:

1) The unit test accurately verifies a segfault for the vulnerable code ✅
2) The header code patch works as intended for all unit tests and my integration tests ✅
3) I have staged the patch in my local repo and I am awaiting approval from our OSS program office.

Thanks for figuring out the segfault unit test. I lacked the Lua/NGiNX test framework knowledge to verify a crashed worker process.
